src/Dscp/Educator/Validate.hs

```diff
@@ -2,34 +2,67 @@
 -- | Validation of private block
 
 module Dscp.Educator.Validate
-       ( validatePrivateBlk
+       ( ValidationFailure (..)
+       , validatePrivateBlk
        ) where
 
 import Universum
 
-import Serokell.Util.Verify (verResToMonadError, verifyGeneric)
+import Data.Text.Buildable (build)
 
 import Dscp.Core (Address (..), SignedSubmission (..), Submission (..),
-                  SubmissionWitness (..))
-import Dscp.Crypto (fromFoldable, getMerkleRoot, hash, verify)
+                  SubmissionSig, SubmissionWitness (..))
+import Dscp.Crypto (Hash, PublicKey, MerkleSignature, fromFoldable, getMerkleRoot, hash, verify)
 import Dscp.Educator.Block (PrivateBlock (..), PrivateBlockHeader (..), pbBody, pbbTxs)
 import Dscp.Educator.Serialise ()
 import Dscp.Educator.Txs (PrivateTx (..))
 
+import qualified Text.Show
+import qualified Data.Text.Buildable ()
+
+
+-- | Validation failures
+data ValidationFailure
+    = MerkleSignatureMismatch !(MerkleSignature PrivateTx) !(MerkleSignature PrivateTx)
+      -- ^ Header merkle tree root does not match
+      -- body merkle tree root
+    | TxPublicKeyMismatch !(Hash PublicKey) !(Hash PublicKey)
```

Comment:
It's better to semantically separate _block header_ validation errors and _transaction_ validation errors. Better to do like this:
```
data BlockValidationFailure
    = MerkleSignatureMismatch ....
    | TxInvalid !TxValidationFailure
    ...

data TxValidationFailure
    = TxPublicKeyMismatch ...
    | TxSignatureMismatch ...
```
Also, it's better to rename `TxValidationFailure` to `SubmissionValidationFailure`, because we're actually checking submission signatures here. These both will allow to re-use `SubmissionValidationFailure` in API logic of validation incoming submissions.

Reply:
Also, please, use record syntax with proper field names, like:
```
    ...
    = MerkleSignatureMismatch 
        { bvfExpectedMerkleSig :: !(MerkleSignature PrivateTx)
        , bvfActualMerkleSig   :: !(MerkleSignature PrivateTx) }
    ...
```